src/reconcile/rec_visibility.c

```diff
@@ -265,64 +256,54 @@ static int
     /*
-     * If we're in preserve_prepared mode and have a prepared time window, look for the prepared
-     * update that triggered the history store deletion. This helps us determine if we should delay
-     * the deletion until the prepared update is committed/rolled back and stable.
+     * If we select an update in the history store to write to disk, delete it from the history
+     * store.
      */
-    for (delete_upd = upd_select->tombstone != NULL ? upd_select->tombstone : upd_select->upd,
-        triggering_prepare_upd = NULL;
-         delete_upd != NULL; delete_upd = delete_upd->next) {
-        if (find_triggering_prepare && delete_upd->type != WT_UPDATE_RESERVE) {
-            WT_ACQUIRE_READ(prepare_state, delete_upd->prepare_state);
-            if (prepare_state != WT_PREPARE_INIT)
-                triggering_prepare_upd = delete_upd;
+    if (F_ISSET(upd_select->upd, WT_UPDATE_HS | WT_UPDATE_HS_MAX_STOP)) {
+        if (upd_select->tombstone == NULL)
+            tombstone = NULL;
+        else if (F_ISSET(upd_select->tombstone, WT_UPDATE_HS))
+            tombstone = upd_select->tombstone;
+        else {
+            tombstone = NULL;
+            WT_ASSERT(session, !F_ISSET(upd_select->tombstone, WT_UPDATE_HS_MAX_STOP));
         }
+        WT_RET(__rec_delete_hs_upd_save(session, r, ins, rip, upd_select->upd, tombstone));
+        return (0);
+    }
 
-        if (delete_upd->txnid == WT_TXN_ABORTED)
+    seen_committed = !WT_TIME_WINDOW_HAS_START_PREPARE(&upd_select->tw);
+
+    /* Delete all the history store record with a max stop timestamp. */
+    for (delete_upd = upd_select->upd->next; delete_upd != NULL; delete_upd = delete_upd->next) {
+        if ((txnid = delete_upd->txnid) == WT_TXN_ABORTED)
             continue;
 
-        if (F_ISSET(delete_upd, WT_UPDATE_TO_DELETE_FROM_HS)) {
-            WT_ASSERT_ALWAYS(session,
-              F_ISSET(delete_upd, WT_UPDATE_HS | WT_UPDATE_RESTORED_FROM_HS),
-              "Attempting to remove an update from the history store in WiredTiger, but the "
-              "update was missing.");
-            find_triggering_prepare = false;
-            if (triggering_prepare_upd != NULL) {
-                uint64_t txnid;
-                WT_ACQUIRE_READ(txnid, triggering_prepare_upd->txnid);
-                if (txnid == WT_TXN_ABORTED)
-                    txnid = triggering_prepare_upd->upd_saved_txnid;
-                /*
-                 * It must be the operation on this prepared update that triggered us to delete the
-                 * update from the history store. If it is the same prepared update we are writing
-                 * to disk, don't delete the update from the history store now. We need to wait the
-                 * prepared update to become stable.
-                 */
-                if (txnid == upd_select->tw.start_txn)
-                    break;
-            }
+        if (!seen_committed && upd_select->tw.start_ts != txnid)
```

Comment:
this should be `upd_select->tw.start_txn != txnid`?

Reply:
good catch. thank you